stl/inc/ranges

```diff
@@ -2260,6 +2214,416 @@ namespace ranges {
+                    }
+                }
+                if constexpr (_Deref_is_glvalue) {
+                    _Inner = _InnerIter{};
+                }
+            }
+
+#if _ITERATOR_DEBUG_LEVEL != 0
+            constexpr void _Check_dereference() const noexcept {
+                _STL_VERIFY(_Parent != nullptr, "cannot dereference value-initialized join_view iterator");
+                _STL_VERIFY(_Outer != _RANGES end(_Parent->_Range), "cannot dereference end join_view iterator");
+                sentinel_t<_InnerRng> _Last;
+                if constexpr (_Deref_is_glvalue) {
+                    _Last = _RANGES end(*_Outer);
+                } else {
+                    _Last = _RANGES end(_Parent->_Inner);
+                }
+                _STL_VERIFY(_Inner != _Last, "cannot dereference end join_view iterator");
+            }
+
+            constexpr void _Same_range(const _Iterator& _Right) const noexcept {
+                _STL_VERIFY(_Parent == _Right._Parent, "cannot compare incompatible join_view iterators");
+            }
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+
+        public:
+            using iterator_concept =
+                conditional_t<_Deref_is_glvalue && bidirectional_range<_Base> && bidirectional_range<_SubRng>,
+                    bidirectional_iterator_tag,
+                    conditional_t<_Deref_is_glvalue && forward_range<_Base> && forward_range<_SubRng>,
+                        forward_iterator_tag, input_iterator_tag>>;
+            using value_type      = range_value_t<_SubRng>;
+            using difference_type = common_type_t<range_difference_t<_Base>, range_difference_t<_SubRng>>;
+
+            _Iterator() = default;
+
+            constexpr _Iterator(_Parent_t& _Parent_, _OuterIter _Outer_)
+                : _Outer{_STD move(_Outer_)}, _Parent{_STD addressof(_Parent_)} {
+#if _ITERATOR_DEBUG_LEVEL != 0
+                _Adl_verify_range(_Outer, _RANGES end(_Parent_._Range));
+                if constexpr (forward_range<_Base>) {
+                    _Adl_verify_range(_RANGES begin(_Parent_._Range), _Outer);
+                }
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+                _Satisfy();
+            }
+
+            // clang-format off
+            constexpr _Iterator(_Iterator<!_Const> _It)
+                requires _Const && convertible_to<iterator_t<_Vw>, _OuterIter>
+                      && convertible_to<iterator_t<_InnerRng>, _InnerIter>
+                : _Outer{_STD move(_It._Outer)}, _Inner{_STD move(_It._Inner)}, _Parent{_It._Parent} {}
+            // clang-format on
+
+            _NODISCARD constexpr decltype(auto) operator*() const noexcept(noexcept(*_Inner)) /* strengthened */ {
+#if _ITERATOR_DEBUG_LEVEL != 0
+                _Check_dereference();
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+                return *_Inner;
+            }
+
+            // clang-format off
+            _NODISCARD constexpr _InnerIter operator->() const // Per resolution of LWG-3500

```

Comment:
LWG-3500 "`join_view::iterator::operator->()` is bogus" was accepted at the February 2021 virtual plenary meeting, so any comments citing it should be removed. :tada: